### DISCUSSION TITLE
Ability to specify the desired dtype when decoding a variable with scale_factor and add_offset

### DISCUSSION THREAD

**Discussion by eschalkargans** [pydata/xarray#9254] [Q&A]
Hello,

I create this discussion following a recent xarray update that changes the decoding of encoded data via attributes `scale_factor` and `add_offset`. Long story short, before, my `uint16` was decoded to a `float32`, but now is decoded to `float64`. The `float32` was desirable for my usecase.

More details on this comment I made on the  issue: https://github.com/pydata/xarray/issues/2304#issuecomment-2230867198 

`_choose_float_dtype` implementation in the associated PR: https://github.com/pydata/xarray/pull/8713/files#diff-ebbe6dc9f2f17dc4a9777193c923783413a3a6814f6e85d82ade7ca9f8ca65c9R396

My question would be: is it possible to specify the desired target dtype? Either via a special encoding attribute (like `scale_factor` and `add_offset`, we could imagine `encoded_dtype='uint16'` and `decoded_dtype='float32'`? Or via a special kwarg to pass to `open_dataset`, like `open_dataset(decode_cf=True, decoding_mapping={'/path/to/my_variable': 'float32'}`?

Thanks!

**Comment by kmuehlbauer** [comment]
@eschalkargans 

The desired target dtype can be specified by setting `scale_factor` and `add_offset` to this wanted dtype. 

The CF conventions explicitly mention this here http://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#packed-data

_When packed data is written, the scale_factor and add_offset attributes must be of the same type as the unpacked data, which must be either float or double. Data of type float must be packed into one of these types: byte, unsigned byte, short, unsigned short. Data of type double must be packed into one of these types: byte, unsigned byte, short, unsigned short, int, unsigned int.

When packed data is read, it should be unpacked to the type of the scale_factor and add_offset attributes, which must have the same type if both are present. For guidance only, we suggest that packed data which does not conform to the rules of this section regarding the types of the data variable and attributes should be unpacked to double type, in order to minimise the risk of loss of precision._

**Reply by eschalkargans** [reply]
The issue is that the granularity allowed by the two abstract data types, floating or integer is not enough to fully describe the desired target dtype. Indeed, float32 and float64 are both floating, and the use of one or the other is system-dependant (`np.dtype(type(float))`). The previous behaviour was to use float32.The function `_choose_float_dtype` contains this logic. (See diff: https://github.com/pydata/xarray/pull/8713/files?diff=unified&w=0#diff-ebbe6dc9f2f17dc4a9777193c923783413a3a6814f6e85d82ade7ca9f8ca65c9L371 )

Since `scale_factor` and `add_offset` are stored as JSON in Zarr, there is no way to specify the byte count of a number natively in JSON unfortunately. Hence the need to be able to explicitly say the wanted target dtype.

Currently either I need to downcast the loaded array from float64 to float32, or I must `decode_cf=False` (which I would like to avoid as the automatic decoding is practical)

**Reply by keewis** [reply]
what version of `xarray` are you using? And which version of `numpy`? If you have `numpy>=2.0`, you'd have to use `main` (or the upcoming `xarray=2024.07.0`, once that's released), because in `2024.06.0` the decoder is not compatible with `numpy>=2.0`.

**Reply by eschalkargans** [reply]
Excerpts from `xr.show_versions()`: 

xarray: 2024.5.0
pandas: 2.2.2
numpy: 1.26.4

**Reply by eschalkargans** [reply]
@kmuehlbauer 

> The desired target dtype can be specified by setting scale_factor and add_offset to this wanted dtype.

This is not possible when serializing to Zarr, because such information is stored as JSON, and JSON can represent only integers or floatings, without any consideration for the size in bytes. So we cannot distinguish from floats and doubles

**Reply by kmuehlbauer** [reply]
@eschalkargans I totally understand the issue with zarr but isn't that an implementation detail of zarr? Could this by chance be handled in the zarr backend code instead of the CF coder?

IMHO, this isn't CF specific and should be handled in the backend code. That's from my perspective, others might have other opinions, though.

**Reply by dcherian** [reply]
You could disable that part of the decoding pipeline and do it yourself?

**Reply by dcherian** [reply]
Alternatively, we could start handling `valid_min` and `valid_max` and use those to choose the smallest type that can represent data in that range.
